turbine/src/broadcast_stage/standard_broadcast_run.rs

```diff
@@ -238,33 +201,73 @@ impl StandardBroadcastRun {
         inc_new_counter_info!("broadcast_service-entries_received", num_entries);
-        let old_broadcast_start = self.slot_broadcast_start;
-        let old_num_batches = self.num_batches;
-        if self.current_slot_and_parent.is_none()
-            || bank.slot() != self.current_slot_and_parent.unwrap().0
-        {
-            self.slot_broadcast_start = Some(Instant::now());
-            self.num_batches = 0;
-            let slot = bank.slot();
-            let parent_slot = bank.parent_slot();
-
-            self.current_slot_and_parent = Some((slot, parent_slot));
-            receive_elapsed = Duration::new(0, 0);
-            coalesce_elapsed = Duration::new(0, 0);
-        }
 
         let mut process_stats = ProcessShredsStats::default();
 
         let mut to_shreds_time = Measure::start("broadcast_to_shreds");
         let cluster_type = bank.cluster_type();
 
-        // 1) Check if slot was interrupted
-        let prev_slot_shreds = self.finish_prev_slot(
-            keypair,
-            bank.ticks_per_slot() as u8,
-            cluster_type,
-            &mut process_stats,
-        );
+        if self.slot != bank.slot() {
+            // Finish previous slot if it was interrupted.
+            if !self.completed {
+                let shreds = self.finish_prev_slot(
+                    keypair,
+                    bank.ticks_per_slot() as u8,
+                    cluster_type,
+                    &mut process_stats,
+                );
+                debug_assert!(shreds.iter().all(|shred| shred.slot() == self.slot));
+                // Broadcast shreds for the interrupted slot.
+                let batch_info = Some(BroadcastShredBatchInfo {
+                    slot: self.slot,
+                    num_expected_batches: Some(self.num_batches + 1),
+                    slot_start_ts: self.slot_broadcast_start,
+                    was_interrupted: true,
+                });
+                let shreds = Arc::new(shreds);
+                socket_sender.send((shreds.clone(), batch_info.clone()))?;
+                blockstore_sender.send((shreds, batch_info))?;
+            }
+            // If blockstore already has shreds for this slot,
+            // it should not recreate the slot:
+            // https://github.com/solana-labs/solana/blob/92a0b310c/ledger/src/leader_schedule_cache.rs##L139-L148
+            if blockstore
+                .meta(bank.slot())
+                .unwrap()
+                .filter(|slot_meta| slot_meta.received > 0 || slot_meta.consumed > 0)
+                .is_some()
+            {
+                process_stats.num_extant_slots += 1;
+                // This is a faulty situation that should not happen.
+                // Refrain from generating shreds for the slot.
+                return Err(Error::DuplicateSlotBroadcast(bank.slot()));
+            }
+            // Reinitialize state for this slot.
+            let chained_merkle_root = (self.slot == bank.parent_slot())
```

Comment:
This is the only logical change right?

Reply:
This and maintaining `self.chained_merkle_root` across slots.